Fix partial write case when using _fd_write/doWritev/writev

jonahgraham opened the pull request:
doWritev iterates through the iovs structure writing the data for each entry. The bug fixed here is that if a write does not write all data, the loop continues. If the subsequent write also accepts data, it has accepted data out of order.

Without the fix to doWritev this new test outputs:

```
TO DEVICE: "A"
TO DEVICE: "X"
TO DEVICE: "C"
TO DEVICE: "X"
TO DEVICE: "Y"
TO DEVICE: "Z"
TO DEVICE: "a"
TO DEVICE: "\n"
TO DEVICE: "c"
TO DEVICE: "\n"
done
```

Fixes #22258

sbc100 commented on test/test_core.py:
```diff
@@ -5942,6 +5942,10 @@ def test_fs_writev(self):
     self.do_runf('fs/test_writev.c', 'success')
 
+  def test_fs_writev_partial_write(self):
+    self.set_setting('FORCE_FILESYSTEM')
```
Is this line needed?

jonahgraham replied:
It isn't. I'll remove in my next commit.

sbc100 commented on test/fs/test_writev_partial_write.c:
```diff
@@ -0,0 +1,84 @@
+    fputs("\n", f);
+    fflush(f);
+    fclose(f);
+  }
```
Could you split this function into 3 functions: `setup`, `test_writev_direct`, `test_via_stdio` and then call each of the from main?

jonahgraham replied:
Yes.

sbc100 reviewed (commented):
Thanks do much for the fix and the great test case.

I wonder if its enough to put this test in test_other.py?  I..e maybe we don't need to run it all the different configurations?

jonahgraham commented:
> I wonder if its enough to put this test in test_other.py? I..e maybe we don't need to run it all the different configurations?

I have moved it. I don't think it needs running in multiple configurations. I wasn't really sure where to put the test, please let me know if the new place is good.

I have added the review comments as additional commits, with the expectation that it will be squashed when merged. Let me know if you want me to squash them now.

kripken commented on test/test_other.py:
```diff
@@ -15028,3 +15028,6 @@ def test_extra_struct_info(self):
     self.assertContained('(before) AF_INET=2', stderr)
     self.assertContained('(after) AF_INET=42', stderr)
+
+  def test_fs_writev_partial_write(self):
```
```suggestion
  @also_with_wasmfs
  def test_fs_writev_partial_write(self):
```
This will also test in WasmFS. The fix doesn't affect it, but I think it's worth verifying that we don't have this bug there (some of the WasmFS code was ported from JS).

jonahgraham replied:
I added this, but how do I run the test with wasmfs? This is what I have been doing to run the test until now: `test/runner other.test_fs_writev_partial_write`

kripken replied:
`test/runner other.test_fs_writev_partial_write_wasmfs` should work, that is, that annotation adds a new test like the original but with suffix `_wasmfs`.

jonahgraham replied:
I have now figured it out: `test/runner other.test_fs_writev_partial_write_wasmfs`

However I see that the test fails with `FS.makedev is not a function` error ([link to circleci](https://app.circleci.com/pipelines/github/emscripten-core/emscripten/35922/workflows/c4064c8f-7e08-4abe-8568-7af99a0e7d75/jobs/802019?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary&invite=true#step-111-377742_39))

So I added `FORCE_FILESYSTEM` back in and it leads to a new error `UNREACHABLE executed`

I am trying to debug that now.

jonahgraham replied:
The unreachable is here: https://github.com/emscripten-core/emscripten/blob/8fdaf16b7d84537d441a469e3ab08471c5f0265e/system/lib/wasmfs/js_impl_backend.h#L104

with this backtrace:

```
RuntimeError: unreachable
    at test_writev_partial_write.wasm.wasmfs::handle_unreachable(char const*, char const*, unsigned int) (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[409]:0xffbe)
    at test_writev_partial_write.wasm.wasmfs::JSImplFile::setSize(long long) (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[636]:0x1b6b0)
    at test_writev_partial_write.wasm.wasmfs::DataFile::Handle::setSize(long long) (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[438]:0x11ff6)
    at test_writev_partial_write.wasm.doOpen(wasmfs::path::ParsedParent, int, unsigned int, wasmfs::Backend*, OpenReturnMode) (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[433]:0x11cb1)
    at test_writev_partial_write.wasm.__syscall_openat (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[439]:0x12095)
    at test_writev_partial_write.wasm.fopen (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[44]:0x1a0c)
    at test_writev_partial_write.wasm.test_via_stdio (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[26]:0x10ce)
    at test_writev_partial_write.wasm.__original_main (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[27]:0x11c0)
    at test_writev_partial_write.wasm.main (wasm://wasm/test_writev_partial_write.wasm-00a47e2e:wasm-function[28]:0x11ee)
    at /scratch/emscripten/emscripten/test_writev_partial_write.js:639:12
```

jonahgraham replied:
(It feels like I should be opening another issue on this, but I continue for here in the meantime)

The unreachable is because we are trying to truncate the file (I `fopen`ed it in `"w"` mode). If I open in `"a"` mode so we don't truncate it works fine.

The truncating happens here:

https://github.com/emscripten-core/emscripten/blob/d72d7226f4733af8ff993dec70198cf09a24142d/system/lib/wasmfs/syscalls.cpp#L563-L564

New/additional commit coming in a moment with this change.

jonahgraham replied:
> `test/runner other.test_fs_writev_partial_write_wasmfs` should work, that is, that annotation adds a new test like the original but with suffix `_wasmfs`.

Somehow I blanked on seeing this comment - much appreciated and sorry for overlooking it!

jonahgraham replied:
> This will also test in WasmFS. The fix doesn't affect it, but I think it's worth verifying that we don't have this bug there (some of the WasmFS code was ported from JS).

FWIW the test passes on WasmFS because this case was fixed in WasmFS as part of https://github.com/emscripten-core/emscripten/pull/16895

Specifically the same condition I added exists here:

https://github.com/emscripten-core/emscripten/blob/d72d7226f4733af8ff993dec70198cf09a24142d/system/lib/wasmfs/syscalls.cpp#L159-L164

(I have a PR in a second that fixes the copy-paste error in the comment)

kripken commented on test/fs/test_writev_partial_write.c:
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2022 The Emscripten Authors.  All rights reserved.
```
```suggestion
 * Copyright 2024 The Emscripten Authors.  All rights reserved.
```

kripken commented on src/library_wasi.js:
```diff
@@ -223,6 +223,7 @@ var WasiLibrary = {
       var curr = FS.write(stream, HEAP8, ptr, len, offset);
       if (curr < 0) return -1;
       ret += curr;
+      if (curr < len) break; // no more space to write
```
```suggestion
      if (curr < len) {
        // No more space to write.
        break;
      }
```

kripken replied:
(just a style change to match our conventions)

kripken reviewed (commented):
Thanks! Looks good aside from those comments.

jonahgraham commented:
Thanks for the reviews and feedback. I am starting to get a hang of the test suite now. I'll revisit any more review feedback tomorrow.

kripken reviewed (approved):
Perfect, thanks!
